src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
> What do they expect to happen in such an event?

Ideally it's easy to explain.

* Releasing an already held resource becomes a NOP.
* Outstanding calls to `EnterAsync` transition to the cancelled state.

In other words, from outside the type, `Dispose` is functionally equivalent to a point-in-time cancellation of all unfulfilled and future requests to `EnterAsync`, without invalidating the ones that were already fulfilled.

In practice we can't achieve the latter, so we instead turn `Dispose` itself into a NOP.